Support NULL on left side of semi join

findepi opened the pull request:
- predicate push down for semi join's build side got disabled
- new replication mode "replicate nulls and any row"
- nulls on left side of `IN` are correctly supported now
- there is session flag unlocking old behavior 

Fixes #6991. Related to #7208, #6622

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -29,10 +29,40 @@
+    {
+        NONE {
+            @Override
+            public boolean isNoReplication()
```
`replication.isNoReplication()` reads weird (even paradoxically ;)). Please replace with `replication == NONE`.

findepi replied:
I had `replication == NONE` but then extracted the method. If some day i change the enum to some configurable being, I will surely need to revisit all usage places but I will not have to fix them all.

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -854,22 +855,22 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, Context context)
                     Partitioning filteringPartitioning = source.getProperties().translate(createTranslator(sourceToFiltering)).getNodePartitioning().get();
-                    filteringSource = node.getFilteringSource().accept(this, context.withPreferredProperties(PreferredProperties.partitionedWithNullsReplicated(filteringPartitioning)));
-                    if (!source.getProperties().withReplicatedNulls(true).isNodePartitionedWith(filteringSource.getProperties(), sourceToFiltering::get)) {
+                    filteringSource = node.getFilteringSource().accept(this, context.withPreferredProperties(PreferredProperties.partitionedWithReplication(filteringPartitioning, Replication.REPLICATE_NULLS)));
```
the convention is to `import static`, and I think it'd read better here and the following occurrences

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ActualProperties.java:
```diff
@@ -372,14 +372,14 @@ private boolean isCoordinatorOnly()
+        private boolean isNodePartitionedOn(Collection<Symbol> columns, Set<Symbol> constants, Replication replication)
         {
-            return nodePartitioning.isPresent() && nodePartitioning.get().isPartitionedOn(columns, constants) && this.nullsReplicated == nullsReplicated;
+            return nodePartitioning.isPresent() && nodePartitioning.get().isPartitionedOn(columns, constants) && Objects.equals(this.replication, replication);
```
`this.replication` is `requireNonNull`, so I'd drop the `Objects.equals` for a ` == `.

findepi replied:
I prefer `Objects.equals(this.replication, replication)` to `this.replication.equals(replication)`

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -74,8 +104,10 @@ public PartitioningScheme(
-        this.replicateNulls = replicateNulls;
+        requireNonNull(replication, "replication is null");
+        checkArgument(replication.isCompatibleWithColumns(columns.size()), "Replication %s cannot be used with %s partitioning columns", replication, columns.size());
+        this.replication = replication;
```
`requireNonNull`?

findepi replied:
It's there, 2 lines earlier. It's not in the same line as assignment, because there is also another check in between.

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java:
```diff
@@ -298,14 +310,17 @@ public Page getOutput()
         private final List<Optional<Block>> partitionConstants;
         private final PagesSerde serde;
         private final List<PageBuilder> pageBuilders;
+        private final boolean replicateAny;
```
judging just from the variable name, `replicateAny` would be true if we're supposed to replicate just nulls. Let's name it `replicateFirstRow`, and the enum value too.

ArturGajowy replied:
sure, but `replicateFirstRow` doesn't add ambiguity to the code and is effectively what we're doing. I'm OK with replacing the `firstRow` part with sth less concrete - as long as it's not introducing ambiguity

findepi replied:
This is not "replicate first row". It is actually "replicate first row from every pipeline", but this is an implementation detail I'm not going to expose.

ArturGajowy replied:
maybe `replicateAnything` then? :)

ArturGajowy replied:
I've seen `replicateAnyRow` from @pnowojski - I'm for it

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java:
```diff
@@ -352,6 +368,7 @@ public PartitionedOutputInfo getInfo()
             Page partitionFunctionArgs = getPartitionFunctionArguments(page);
             for (int position = 0; position < page.getPositionCount(); position++) {
                 if (shouldReplicate(page, position)) {
+                    hasAnyBeenReplicated = true;
```
I'd move this to the end of the if, after the replication actually happens

ArturGajowy commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -5054,11 +5055,24 @@ public void testLargeIn()
+        assertQuery("SELECT NULL IN (1, 2, 3)", "SELECT NULL");
+        assertQuery("SELECT NULL IN (SELECT 1)", "SELECT NULL");
+        assertQuery("SELECT NULL IN (SELECT 1 WHERE FALSE)", "SELECT FALSE");
+        assertQuery("SELECT x FROM (VALUES NULL) t(x) WHERE x IN (SELECT 1)", "SELECT 33 WHERE FALSE");
```
`s/"SELECT 33 WHERE FALSE"/"SELECT 1 WHERE FALSE"` ;)

findepi replied:
`33` was intentionally different from `1` -- there is `1` elsewhere in the query

ArturGajowy commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6213,6 +6225,10 @@ public void testSemiJoinLimitPushDown()
     {
+        assertQuery("WITH empty AS (SELECT 1 WHERE FALSE) " +
+                        "SELECT 3 FROM (VALUES 1) WHERE NULL IN (SELECT * FROM empty)",
+                "SELECT 42 WHERE 0 = 1");
```
"SELECT 1 WHERE FALSE", let's not confuse the reader

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -841,27 +841,36 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> cont
-                    Expression rewrittenConjunctOrNull = expressionOrNullSymbols(Predicate.isEqual(node.getFilteringSourceJoinSymbol())).apply(rewrittenConjunct);
-                    filteringSourceConjuncts.add(rewrittenConjunctOrNull);
+            PlanNode rewrittenFilteringSource;
+            if (node.isLegacySemiJoin()) {
```
I'd greatly appreciate a `rewriteFilteringSource()` method, since its code has been identified :) If at all possible :)

ArturGajowy commented on presto-main/src/main/java/com/facebook/presto/operator/ChannelSet.java:
```diff
@@ -58,18 +61,24 @@ public boolean containsNull()
         private final GroupByHash hash;
         private final OperatorContext operatorContext;
         private final Page nullBlockPage;
+        private final LongAdder elements = new LongAdder();
```
The builder seems to be used in a single thread only, so let's use `long` so that nobody thinks otherwise.

pnowojski commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java:
```diff
@@ -840,7 +840,7 @@ public PlanWithProperties visitUnnest(UnnestNode node, Context context)
         public PlanWithProperties visitSemiJoin(SemiJoinNode node, Context context)
         {
-            Replication partitionedSemiJoinReplication = Replication.REPLICATE_NULLS;
+            Replication partitionedSemiJoinReplication = Replication.REPLICATE_NULLS_AND_ANY;
```
? does this change anything? Why there is no change in tests?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SemiJoinNode.java:
```diff
@@ -38,6 +38,7 @@
     private final Optional<Symbol> sourceHashSymbol;
     private final Optional<Symbol> filteringSourceHashSymbol;
     private final Optional<DistributionType> distributionType;
+    private final boolean legacySemiJoin;
```
I don't think legacy semi join is the best name of the property. I think is should rather be `predicatePushdownEnabled`, `ternaryLogic` or `semiJoinOutputNullDistinguishable`. I imagine this option should be set (in the future) during `SemiJoinNode` creation based on analysis if `null` and `false` value distinction is required.

sopel39 replied:
I think the field name should suggest how the operator works, e.g: produces values in binary or ternary logic. So the name could be: `semiJoinOutputNullDistinguishable` or `semiJoinOutputInTernaryLogic`, etc.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/HashSemiJoinOperator.java:
```diff
@@ -45,17 +45,19 @@
         private final SetSupplier setSupplier;
         private final List<Type> probeTypes;
         private final int probeJoinChannel;
+        private final boolean legacySemiJoin;
```
ditto

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java:
```diff
@@ -106,7 +107,8 @@ public PlanNode visitApply(ApplyNode node, RewriteContext<Void> context)
                     Optional.empty(),
-                    Optional.empty()
+                    Optional.empty(),
+                    SystemSessionProperties.isLegacySemiJoin(context.get())
```
Since importance of ternary logic or binary logic could be deduced before `SemiJoinNode` creation (and without looking into session at all) I think it's good approach to store the flag in `SemiJoinNode`

sopel39 replied:
I think property: "if null is distinguishable from false" is more algebraic feature of `SemiJoin` rather than optimization turned on/off by feature toggle. The property can be obtained in various ways (e.g: using analysis) and other optimizers (e.g: predicate pushdown) have to take it into account.

I think you can treat this property like "partitioning" or `IndexJoin` vs `InnerJoin`. `SemiJoin` producing "ternary logic" is different node than semi join producing binary logic

findepi replied:
I don't get. I don't call it "use IN as in SQL 2003 or choose IN as in My Proprietary Spec 2017".

FWIW, if the flag stays, it will be renamed to something like "support-nulls-on-lhs-of-semi-join" (default true).

sopel39 replied:
The idea is that you can think of `SemiJoinNode` with the flag on and off as two kind of `SemiJoinNodes`.
For instance, `ApplyNode` transformation would set the `SemiJoinNode` flag accordingly (e.g: depending if the semi join output is used in projection or filter) and other optimizers should obey the flag (e.g: predicate pushdown). This way you decouple symbolic analysis from other predicate pushdown. This is one of the ways forward if we want to further enhance lhs support null in semi join.

I don't have strong option on it though. You can remove the flag from `SemiJoinNode` altogether

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/ChannelSet.java:
```diff
@@ -58,18 +61,24 @@ public boolean containsNull()
         return containsNull;
     }
 
+    public boolean hasElements()
```
`s/hasElements/isEmpty/g`

and you can implement it as `return size() == 0`

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -29,10 +29,40 @@
 
 public class PartitioningScheme
 {
+    public enum Replication
```
why not also add method: `replicateNulls` instead of using `isReplicateNulls()`. This way it will be more objective and all `PartitioningScheme` have to declare how they replicate data instead of relying on `switch`.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -348,8 +349,9 @@ public LocalExecutionPlan plan(
-        if (partitioningScheme.isReplicateNulls() && partitioningColumns.size() == 1) {
+        boolean isReplicateNulls = isReplicateNulls(partitioningScheme.getReplication());
+        checkArgument(!isReplicateNulls || partitioningColumns.size() <= 1);
+        if (isReplicateNulls && partitioningColumns.size() == 1) {
```
shouldn't you use `isCompatibleWithColumns` here?

findepi replied:
Not sure. `isCompatibleWithColumns` is generic, implementation agnostic. Here you have code assertion ensuring `getOnlyElement(partitioningColumns)` is the right way to do.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -29,10 +29,40 @@
 {
+    public enum Replication
+    {
+        NONE {
```
instead of overriding method in each `Replication` you could use:
```
public enum Replication {
   NONE(true, Integer.MAX_VALUE), REPLICATE_NULLS(false, 1);

   private final boolean noReplication;
   private final int maxReplicationColumns;

   Replication(...) {
   }
   
   public boolean isNoReplication() {...}
   public int isCompatibleWithColumns(...) {...}
   
}
```

findepi replied:
Sure I could. Not sure this would be more readable though.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -123,8 +126,15 @@ public static ActualProperties deriveProperties(PlanNode node, List<ActualProper
+        if (!(node instanceof SemiJoinNode)) {
+            Set<Replication> inputReplications = inputProperties.stream()
+                    .map(ActualProperties::getReplication)
+                    .filter(replication -> !replication.isNoReplication())
```
double negation here. Maybe rename
`isNoReplication` to `replicates`? Then you could write clearer stream:
```
.filter(Replication::replicates)
```

sopel39 replied:
`.filter(Replication::replicates)` is less verbose than:
`.filter(replication -> !replication.isNoReplication())`.
Other places don't seem to be affected either way.
It is also mentally easier to think in terms of direct logic without intermittent negation.
Any logical reason to keep `isNoReplication()`?

findepi replied:
I changed it to `replicatesNothing`. I have no problem with adding `replicatesSomething` to the `Replication`, just not sure it makes API more easier to use.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -123,8 +126,15 @@ public static ActualProperties deriveProperties(PlanNode node, List<ActualProper
+                    .filter(replication -> !replication.isNoReplication())
+                    .collect(toImmutableSet());
+            verify(inputReplications.size() == 0
+                            || (inputReplications.size() == 1 && Objects.equals(output.getReplication(), getOnlyElement(inputReplications))),
```
I don't exactly know context of it, but it seems that either:
* none input should be replicated
* the only one input should be replicated the same as output

Here you have:
some replicates inputs should be replicated the same as output. That is: there might be unreplicated input left.

findepi replied:
Neither do I. It seems we're operating on single element list and making check over all elements... I didn't want to change this place though.

sopel39 replied:
But your code has different semantics than what was before. The closest of what was before is to check that all `inputReplications` are the same as `output.getReplication()`. There was no `inputReplications.size() == 1` condition before

findepi replied:
I check the replication is preserved (as it used to be). The `inputReplications.size() == 1` means "inputs all have same replication" and then I can check output has the same replication. I do _not_ check there was single input or something

pnowojski commented on presto-main/src/main/java/com/facebook/presto/operator/HashSemiJoinOperator.java:
```diff
@@ -180,9 +180,19 @@ public void addInput(Page page)
         // update hashing strategy to use probe cursor
         for (int position = 0; position < page.getPositionCount(); position++) {
             if (probeJoinPage.getBlock(0).isNull(position)) {
```
maybe extract for loop to separate method?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java:
```diff
@@ -435,7 +445,9 @@ public ActualProperties visitIndexSource(IndexSourceNode node, List<ActualProper
         {
-            checkArgument(node.getScope() != REMOTE || inputProperties.stream().noneMatch(ActualProperties::isNullsReplicated), "Null replicated inputs should not be remotely exchanged");
+            checkArgument(node.getScope() != REMOTE ||
+                            inputProperties.stream().map(ActualProperties::getReplication).allMatch(Replication::isNoReplication),
```
`noneMatch(Replication::replicates`)?

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java:
```diff
@@ -238,18 +239,11 @@ private static String formatFragment(Metadata metadata, Session session, PlanFra
-                    Joiner.on(", ").join(arguments),
-                    formatHash(partitioningScheme.getHashColumn())));
-        }
+        builder.append(format("Output partitioning: %s%s [%s]%s\n",
```
please update `explain.rst` doc if output changes

findepi replied:
I didn't find any `(replicate nulls)` to update in the docs.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java:
```diff
@@ -317,6 +332,7 @@ public PagePartitioner(
             this.partitionConstants = requireNonNull(partitionConstants, "partitionConstants is null").stream()
                     .map(constant -> constant.map(NullableValue::asBlock))
                     .collect(toImmutableList());
+            this.replicateAny = requireNonNull(replication, "replication is null") == Replication.REPLICATE_NULLS_AND_ANY;
```
please add `replicateAny()` method to `Replication` enum (wrap enum logic into methods). Then you also wouldn't need `replicateAny` field here

findepi replied:
I'm not convinced this is worth it. It's single use place and the actual enum constant meaning is very... this-place-specific.

sopel39 replied:
Doesn't matter. By doing `==` you export business logic from `Enum` to user classes, which is more fragile design than centralized logic.

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -44,6 +44,16 @@ boolean isCompatibleWithColumns(int columnsCount)
+                return columnsCount <= 1;
+            }
+        },
+
```
is `REPLICATE_NULLS` still required?

findepi replied:
It's not.

sopel39 replied:
Remove it then

sopel39 commented on presto-main/src/main/java/com/facebook/presto/operator/ChannelSet.java:
```diff
@@ -29,12 +30,14 @@
 {
     private final GroupByHash hash;
     private final boolean containsNull;
+    private final long elements;
```
you can just have `isEmpty` flag. `elements` suggest that this number is somehow reliable (as per total number of elements in cluster). What you are interested is only if the set is empty (on whole cluster).

sopel39 commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -5054,11 +5055,24 @@ public void testLargeIn()
+    }
+
+    @Test
+    public void testNullOnLhsOfInPredicateAllowed()
```
bunch of tests that should pass:
```
        // test "null IN" subqueries
        assertQuery("SELECT null IN (SELECT cast(NULL AS BIGINT))");
        assertQuery("SELECT null IN (SELECT null where false)");
        assertQuery("SELECT null IN (SELECT 1)");
        assertQuery("SELECT null IN (SELECT 1 where false)");
        assertQuery("select null in ((select 1) union all (select null))");
        assertQuery("select x in (select true) from (select * from (values cast(null as boolean)) t(x) where (x or null) is null)");
        assertQuery("select x in (select 1) from (select * from (values cast(null as integer)) t(x) where (x + 10 is null) or x = 2)");
        assertQuery("select x in (select 1) from (select * from (values cast(null as integer)) t(x) where (x + 10 is null) or x = 1)");
        assertQuery("select x in (select true) from (select * from (values cast(null as boolean)) t(x) where x is null)");
```

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java:
```diff
@@ -61,14 +62,14 @@
      */
     private static class InPredicateRewriter
-            extends SimplePlanRewriter<Void>
+            extends SimplePlanRewriter<Session>
```
You don't have to pass session via context at all. Yo can just make it field in `InPredicateRewriter`

findepi replied:
commit dropped altogether

findepi commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -144,6 +145,18 @@ public boolean isLegacyMapSubscript()
         return legacyMapSubscript;
     }
 
+    @Config("deprecated.legacy-semi-join")
```
Because I don't want to support it for ever

findepi commented on presto-main/src/main/java/com/facebook/presto/operator/HashSemiJoinOperator.java:
```diff
@@ -45,17 +45,19 @@
             this.probeTypes = ImmutableList.copyOf(probeTypes);
             checkArgument(probeJoinChannel >= 0, "probeJoinChannel is negative");
             this.probeJoinChannel = probeJoinChannel;
+            this.legacySemiJoin = legacySemiJoin;
```
I ended up setting the flag in `HashSemiJoinOperator`.  Naturally, I don't want to eval the flag at every `page`.

ArturGajowy commented:
One more thing I forgot - the docs need an update. Search for `This limitation is needed`, there should be 2 places :)

sopel39 commented:
For the future, please add fixups commits and let review them before squashing. This way it's easier to track what has changed.

findepi commented:
I couldn't not squash the fixup. There were too many and they would in no way apply. Sorry for causing more review trouble.

ArturGajowy commented:
You might want to document the feature flag, but I don't insist until we know if it's there to stay :)

findepi commented:
@sopel39 applied

findepi commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -284,6 +285,11 @@ public SystemSessionProperties(
+                        SUPPORT_NULL_ON_LHS_OF_SEMI_JOIN,
+                        "Support NULL on left-hand side of Semi Join (IN), at the cost if disabling certain optimizations",
+                        featuresConfig.isSupportNullOnLhsOfSemiJoin(),
+                        true),
```
Should not be hidden

martint commented on presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:
```diff
@@ -283,6 +284,11 @@ public SystemSessionProperties(
                         featuresConfig.isLegacyOrderBy(),
                         false),
                 booleanSessionProperty(
+                        SUPPORT_NULL_ON_LHS_OF_SEMI_JOIN,
```
I don't think we need a feature toggle. If I recall correctly, the "optimization" was not only incorrect in the general case, but it was almost useless in practice.

sopel39 replied:
It's not obvious for me that it is almost useless in practice. You can easily come up with query where predicate pushdown will make a great difference for semi join, e.g:
```
SELECT x IN (SELECT y FROM very_large_table) FROM ... WHERE x=3
```

Why it would be almost useless in practice for semi join and useful for inner join for example?

findepi replied:
@martint as i promised i run tpcds without and with this change. For the queries containing `IN` (15 of them, including `qNN_N` variants) no performance regression was observed.

Still, maybe we want the switch temporarily, to be on the safe side?

@KBP-TDC any thought?

findepi replied:
@sopel39 in your example, can you somehow push the predicate down _correctly_? Maybe if you knew `x` cannot be null, then `x IN (select y from very_large_table where y=3 or y is null)`, but currently we don't have nullability  in our type system.

sopel39 replied:
@findepi I think for example I provided it is pushed correctly, see: `expressionOrNullSymbols` method

martint replied:
> It's not obvious for me that it is almost useless in practice. You can ...

Fair enough.

> I think for example I provided it is pushed correctly,

But, it it's correct, then what's the issue? It's been a while since I looked at this code and had a conversation about it, so I'm forgetting the details.

sopel39 replied:
I think @findepi was concerned that PP won't push `y=3 or y is null` but only `y=3`

findepi replied:
IIRC the code will push `y=3 or y is null` but it is correct *only if* we know `x` is not null. Which we don't currently know. Otherwise there wouldn't be any correctness issue nor this PR :)

martint replied:
If `x = 3` (because of the predicate) then we know it's not null.

findepi replied:
Anyway, I dropped the feature toggle and updated implementation accordingly (hence rebase...). I also added a TODO in `EffectivePredicateExtractor` where it is incorrect (projections and nulls). There was an existing TODO, but it was too brief to be clear.

Maybe fixing `EffectivePredicateExtractor` is what it takes to re-enable certain pushdowns.

@martint  back to you.

sopel39 replied:
> I also added a TODO in EffectivePredicateExtractor where it is incorrect (projections and nulls)

AFAIR predicates extracted from `EffectivePredicateExtractor` are only used for equality inference. Equality inference determines if it is possible to replace one symbol with some equivalence. In this context `EffectivePredicateExtractor` works correctly. Predicate `a=null` extracted from projection means it is is possible to use `null` in all places where `a` would be used (e.g: `f(a)` can be rewritten as `f(null)`), because those values will always be the same in row context. You can think of more complex examples, e.g: projection `a=b+3`. In this case it also works, when `b` is `null`, `a` is `null` too. Because of that I think `EffectivePredicateExtractor` name is misleading. I think it should actually be called `EffectiveEquvalenceExtractor`.

If you fixed the projection as you mentioned, then it would basically break equality inference for predicated derived from projections. This is because you would have to append `or x is null` whenever you are not sure that the symbol value won't be `null`, (e.g: predicate derived from projection `a=b+3` would be `a=b+3 or (a is null)`). Since equality inference doesn't process alternatives it wouldn't use those `ORed` predicates.
Probably we need two mechanisms:
1. `EffectiveEquvalenceExtractor` which is basically current `EffectivePredicateExtractor` (probably more explicit)
2. Some other mechanism to determine nullability of a symbol. Maybe nullability should be a kind of trait. I've seen that calcite has nullability as type parameter. There is also basic nullability analyzer for expressions in Presto (`NullabilityAnalyzer`).

> Maybe fixing EffectivePredicateExtractor is what it takes to re-enable certain pushdowns.

This won't solve all issues, e.g:
`SELECT * FROM ... WHERE (x = 3 OR x is null) AND (x IN (...))`
It doesn't matter that 'x' can be null. It doesn't matter since `IN` is used in filter.

On the other hand:
`SELECT (x IN (...)) FROM ... WHERE (x = 3)`
without nullability analysis we won't push predicate down to semi join.

So we probably need both nullability analysis and null invariance analysis.

sopel39 replied:
> no performance regression was observed.

@findepi  What about peak memory usage?

findepi replied:
> What about peak memory usage?

I didn't measure that. 

> AFAIR predicates extracted from EffectivePredicateExtractor are only used for equality inference. Equality inference determines if it is possible to replace one symbol with some equivalence. [...]

I see the point. Sort of. But I'm not sure this is the whole picture.

For a condition `x = c` `EffectivePredicateExtractor` states `x = c` and we know this _holds for every `x`_ regardless of `c`. (If `c` is null we know `x = c` _holds for **every** `x`_, so there is no `x`).
I _think_ we use this knowledge when handling joins.

For a projection `c as x` we cannot state `x = c` _holds for every `x`_, but `EffectivePredicateExtractor` returns the same output.

I think both kinds of information are valuable:
- (from projection) equivalence: "you can replace `x` with `c` whenever it occurs"
- (from filter) effective predicate: "`x` is equal to `c` whenever occurs"

Perhaps in JOIN both information kinds are useful and in SEMI JOIN only the latter (effective predicate) is?

sopel39 replied:
> I didn't measure that.
Please do, because memory might the difference that is more visible than performance difference.

> I think we use this knowledge when handling joins.

If that would be the case then there would be a bug in join handling because of projections.

> I think both kinds of information are valuable:

Sure they are, but they have to be separated into two kind of extractors.

> Perhaps in JOIN both information kinds are useful and in SEMI JOIN only the latter (effective predicate) is?

Still if you get predicate `x=3 or x is null` you don't want to give up on PP completely, but you can PP predicate like:
`z=3 or pass_first_row()` or similar

sopel39 replied:
> I didn't measure that.

You can do the measurements on local scale, e.g: sf1. % difference matters

findepi replied:
i measured but no clear change observed (small +- but generally random)

sopel39 commented:
> I didn't measure that.

You can do the measurements on local scale, e.g: `sf1`. % difference matters

martint commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java:
```diff
@@ -391,6 +381,16 @@ private Page getPartitionFunctionArguments(Page page)
             return new Page(page.getPositionCount(), blocks);
         }
 
+        private void addToPageBuilder(PageBuilder pageBuilder, Page page, int position)
```
call this `appendRow`

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -22,17 +22,54 @@
 {
+    public enum Replication
+    {
+        REPLICATE_NOTHING(
```
Do these introduce a new concept or just formalize existing ones?

martint replied:
Also, how does this relate to https://github.com/prestodb/presto/pull/5323? (Where we try to capture the notion of a value that is not equal to itself for the purpose of deciding whether to replicate it or not)

martint replied:
The more I think about it, the more convinced I am that this commit is unnecessary (at least, for the purpose of fixing the "nulls on left side of semijoin issue"). At the end of this PR, we still have two possible values for the enum (NOTHING vs NULLS_AND_ANY). We could just rename the `replicateNulls` variable and be done with it.

martint commented on presto-main/src/main/java/com/facebook/presto/operator/HashSemiJoinOperator.java:
```diff
@@ -174,21 +174,8 @@ public void addInput(Page page)
         Page probeJoinPage = new Page(page.getBlock(probeJoinChannel));
 
         // update hashing strategy to use probe cursor
-        for (int position = 0; position < page.getPositionCount(); position++) {
```
What's the purpose of this change?

findepi replied:
Readability, since the loop body is going to grow further.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -877,30 +874,6 @@ public PlanNode visitSemiJoin(SemiJoinNode node, RewriteContext<Expression> cont
             return output;
         }
 
-        private PlanNode rewriteSemiJoinFilteringSource(SemiJoinNode node, RewriteContext<Expression> context, Expression inheritedPredicate, Expression sourceEffectivePredicate)
```
This method was extracted a few commits above, which is kind of pointless since it's being removed here. Squash it with this commit and do the removal in one step.

findepi replied:
It was for review-ability. Squashed.

martint commented on presto-main/src/test/java/com/facebook/presto/operator/TestHashSemiJoinOperator.java:
```diff
@@ -175,8 +175,7 @@ public void testBuildSideNulls(boolean hashEnabled)
         OperatorAssertion.assertOperatorEquals(joinOperatorFactory, driverContext, probeInput, expected, hashEnabled, ImmutableList.of(probeTypes.size()));
```
Squash this commit with the one that fixes the issue.

findepi replied:
Done

martint commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestPredicatePushdown.java:
```diff
@@ -23,6 +23,7 @@
 import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.equiJoinClause;
```
Squash this commit with the one that fixes the issue

findepi replied:
Done

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -41,11 +41,6 @@
         REPLICATE_NULLS_AND_ANY(
                 true,
                 true,
-                columnsCount -> columnsCount <= 1),
```
Squash this commit with the one that removes all usages of REPLICATE_NULLS.

findepi replied:
Done

martint commented:
> Readability, since the loop body is going to grow further.

Only a couple of lines? I'm not convinced it's worth it.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PartitioningScheme.java:
```diff
@@ -22,17 +22,54 @@
+                true,
+                columnsCount -> columnsCount <= 1);
+
+        private final boolean replicatesNulls;
```
I'm thinking this is a bit overkill, especially with the "replicatesAny" that gets added later. I'm not convinced generalizing these aspects of replication makes things easier or more readable (vs just checking against the enum value wherever it matters). I'll play around with the code a bit and see how it ends up.

martint commented on presto-main/src/main/java/com/facebook/presto/operator/PartitionedOutputOperator.java:
```diff
@@ -355,6 +371,7 @@ public PartitionedOutputInfo getInfo()
                     for (PageBuilder pageBuilder : pageBuilders) {
                         appendRow(pageBuilder, page, position);
                     }
+                    hasAnyRowBeenReplicated = true;
```
This doesn't look correct. If `shouldReplicate` returned true due to a null, this will cause none of the non-null rows to be replicated.

Also, I don't like asymmetry and apparent side effect setting this has on `shouldReplicate`. I think it'd be easier to reason about if the code were inlined.

findepi commented:
@martint, as requested, I dropped `Replication` enum and replaced that commit with "Replace replicateNulls with replicateNullsAndAny in partitioning". Still separate not to mix repetitive changes with interesting ones.

findepi commented:
Thanks for review!
